proposals/record-structs.md

```diff
@@ -0,0 +1,301 @@
+# Record structs
+
+The syntax for a record struct is as follows:
+
+```antlr
+struct_record_declaration
+    : attributes? struct_modifier* 'partial'? 'record' 'struct' identifier type_parameter_list?
+      parameter_list? struct_interfaces? type_parameter_constraints_clause* record_struct_body
+    ;
+
+record_struct_body
+    : struct_body
+    | ';'
+    ;
+```
+
+Record struct types are value types, like other struct types. They implicitly inherit from the class `System.ValueType`.
+The modifiers and members of a record struct are subject to the same restrictions as those of structs
+(accessibility on type, modifiers on members, parameterless instance constructors,
+`base(...)` instance constructor initializers, definite assignment for `this` in constructor, destructors, ...).
+
+See https://github.com/dotnet/csharplang/blob/master/spec/structs.md
+
+But instance field declarations for a record struct are permitted to include variable initializers when there is a primary constructor.
+(TODO: need to confirm, since default values still possible)
+
+At most one partial type declaration of a partial record struct may provide a `parameter_list`.
+The `parameter_list` may not be empty.
+
+Record struct parameters cannot use `ref`, `out` or `this` modifiers (but `in` and `params` are allowed).
+
+## Members of a record struct
+
+In addition to the members declared in the record struct body, a record struct type has additional synthesized members.
+Members are synthesized unless a member with a "matching" signature is declared in the record struct body or
+an accessible concrete non-virtual member with a "matching" signature is inherited.
+Two members are considered matching if they have the same
+signature or would be considered "hiding" in an inheritance scenario.
+It is an error for a member of a record struct to be named "Clone".
+It is an error for an instance field of a record struct to have an unsafe type.
+
+A record struct is not permitted to declare a destructor.
+
+The synthesized members are as follows:
+
+### Equality members
+
+The synthesized equality members are similar as in a record class (`Equals` for this type, `Equals` for `object` type, `==` and `!=` operators for this type),\
```

Comment:
I thought that the intent was _not_ to synthesize an equals implementation, but instead work with the runtime to improve the implementation for all structs? https://github.com/dotnet/csharplang/blob/master/meetings/2020/LDM-2020-09-30.md#struct-equality covers this: we wanted being a record to only opt into _new_ functionality such as `IEquatable<T>`.

Reply:
Much of this section seems incorrect to me based on that understanding.

Reply:
The way I understand the notes is that `record struct` is a way of opting into `IEquatable` and also strongly-typed `Equals`. It is an all-or-nothing opt-in.
```
Next, we looked at whether we should expose new equality operators and strongly-typed Equals methods on struct types, as well as implementing IEquatable<T>. We again came to the conclusion that, for all existing struct types, it would be too costly in metadata (and a potential breaking change for exposing a strongly-typed Equals method or IEquatable<T>) to do this for all types. However, we do think that a gesture for opting into this generation would be useful. Given that, we considered whether it was useful to have these be more granular gestures, ie if a type could just opt-in to generating IEquatable<T> without the equality features. For these scenarios, we feel that the need just isn't there, and that it should be an all-or-nothing opt-in.
```

Did I misunderstand?

---
In reply to: [553000880](https://github.com/dotnet/csharplang/pull/4307#discussion_r553000880) [](ancestors = 553000880)

Reply:
It's not how I understood the result of that meeting and not how I intended the notes to be written :). Making it a record exposes more surface area, but the surface area should all still be forwarding back the built-in struct equality functionality and not be synthesizing its own handling (or at least, that's what I took from our debates).